### PR TITLE
Split List into NodeList & ArrayList

### PR DESCRIPTION
**Pull request by atifaziz**
This addresses issue #95 and some feedback from PR #68.

PR #68 added `List<T>` but trouble was that `T` could be constrained to `INode` in order to warrant calling it by another name like `AstList` or `NodeList`. This PR reviews and fixes several relationships such that `T` can be constrained to `INode`. This wasn't done willy-nilly, but rather with consideration to how Esprima also defines those relationships. `List` has therefore been renamed to `NodeList` and `T` is now constrained to `INode`. Another (good) side-effect of this change was that `NodeList<T>` is now purely read-only. Previously, it had internal methods for modification but those parts have now been  moved to `ArrayList<T>` where `T` is unconstrained and so can be generically used for any type. Internally, `ArrayList<T>` is used to build an array and then the ownership yielded to `NodeList<T>` for public exposure as a read-only list. During the transfer, the `T` gets constrained to nodes.

Both `ArrayList<>` and `NodeList<>` are value types so there is no change to the allocation dynamics with this design and split.

## Additional Notes

`ArrayList<>` can be tricky to use so I have introduced testing of the debug build and added some checks in the debug for potentially incorrect use. Having a struct intended for modification can introduce some very subtle and ugly bugs if not used carefully. For example, two copies of the struct start with the same base array and modifying either list may also modify the other. Consider the following:

```c#
var a = new ArrayList<int>();
a.Add(1);
a.Add(2);
a.Add(3);
var b = a;
b.Add(4);
b.RemoveAt(0);
```

Both `a` and `b` will see the same changes. However, they'll appear to change independently if the example is changed as follows:

```c#
var a = new ArrayList<int>();
a.Add(1);
a.Add(2);
a.Add(3);
var b = a;
b.Add(4);
b.Add(5);        // <-- only new change
b.RemoveAt(0);
```

When 5 is added to `b`, `b` re-allocates its array to make space and consequently further changes are only visible in `b`. To help avoid these subtle bugs, the debug version of this implementation tracks changes. It maintains a local and a boxed version number.The boxed version gets shared by all copies of the struct. If a modification is made via any copy then the boxed version number is updated. Any subsequent use (even if for reading only) of other copies check that their local version numbers haven't diverged from the shared one. In effect, if a copy is made and modified then the original will throw if ever used. For the example above, this means while it's safe to continue to use copy `b` after
modification, `a` will become useless:

```c#
var a = new ArrayList<int>();
a.Add(1);
a.Add(2);
a.Add(3);
var b = a;
b.Add(4);
b.Add(5);
b.RemoveAt(0);
Console.WriteLine(b.Count); // safe to continue to use
Console.WriteLine(a.Count); // will throw
```

### CODE REVIEW

File: src/Esprima/Ast/ArrowFunctionExpression.cs [hunk 1]
```diff
@@ -4,21 +4,15 @@ namespace Esprima.Ast
 {
     public class ArrowFunctionExpression : Node, Expression, IFunction
     {
-        public bool Strict { get; }
-
         public Identifier Id { get; }
-
-        public List<INode> Params { get; }
-
+        public NodeList<INode> Params { get; }
```
Comment (lahma) [inline src/Esprima/Ast/ArrowFunctionExpression.cs]: I don't know as my knowledge is still a bit lacking, could we have sort of ref-readonly solution

File: appveyor.yml [hunk 2]
```diff
@@ -10,6 +10,7 @@ build_script:
   - dotnet --version
   - dotnet pack -c Release
 test_script:
+  - dotnet test .\test\Esprima.Tests\Esprima.Tests.csproj -c Debug -f netcoreapp2.1
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply appveyor.yml]: Because it will catch misuses of `ArrayList<>` that the release version won't.

File: src/Esprima/ArrayList.cs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        //     a.Add(2);
+        //     a.Add(3);
+        //     var b = a;
+        //     b.Add(4);
+        //     b.Add(5);
+        //     b.RemoveAt(0);
+        //     Console.WriteLine(b.Count); // safe to continue to use
+        //     Console.WriteLine(a.Count); // will throw
+
+#if DEBUG
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply src/Esprima/ArrayList.cs]: Right.

File: src/Esprima/Ast/ArrayExpression.cs [hunk 4]
```diff
@@ -4,9 +4,9 @@ namespace Esprima.Ast
 {
     public class ArrayExpression : Node, Expression
     {
-        public readonly List<ArrayExpressionElement> Elements;
+        public readonly NodeList<ArrayExpressionElement> Elements;
 
-        public ArrayExpression(List<ArrayExpressionElement> elements) :
+        public ArrayExpression(NodeList<ArrayExpressionElement> elements) :
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply src/Esprima/Ast/ArrayExpression.cs]: Why comment on a deleted line?

> not it's a copy unless a ref, do people need to construct manually

I can't quite parse this. Could you rephrase and/or expand, please?

[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Esprima/Ast/ArrayExpression.cs]: I think the concern here is that when the `ArrayExpression` is constructed, the `elements` collection is passed by value, cloning the whole array.

File: src/Esprima/Ast/ArrowFunctionExpression.cs (hunk 1 above)
Comment (sebastienros) [reply src/Esprima/Ast/ArrowFunctionExpression.cs]: I looked at the `JavaScript` parser and when the constructors are called with `Params` or `elements` it's once these collection are already built. Won't it just reallocate all its elements then only to have the collection as a struct? Where is the gain then?

File: src/Esprima/Ast/ArrayExpression.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/Ast/ArrayExpression.cs]: tried to fix a typo, but Sebastien pretty much coined it. I think the struct is copied, even though the referenced array might be intact and happily in heap, only pointer with the struct.

[the remark this answers is not in the document]
Comment (atifaziz) [reply src/Esprima/Ast/ArrayExpression.cs]: Ah, I see the point being raised here.

> cloning the whole array

It doesn't clone the whole array. It (_logically_) copies the structure, which is the _array reference_ plus the count. However, this situation is no different with `NodeList<>` than it was with `List<>` previously. Micro-optimisation concerns should be raised as a separate issue and addressed in a separate PR to avoid bundling too many things in one review. Don't you think?

File: src/Esprima/Ast/ArrowFunctionExpression.cs (hunk 1 above)
Comment (atifaziz) [reply src/Esprima/Ast/ArrowFunctionExpression.cs]: @lahma Yes, that could be done in a separate step and it would have to go on the interface! I just don't think that's something we need to address immediately in this PR. The situation is no different than with `List<>`.

@sebastienros There won't be any new allocation. During parsing, the collection/list is built up using `ArrayList<>`, which is just a thin (`struct`) wrapped around an array with an element count. When a node requiring such a list is constructed, the array ownership is _transferred_ from `ArrayList<>` to `NodeList<>`.

Comment (atifaziz) [reply src/Esprima/Ast/ArrowFunctionExpression.cs]: > Where is the gain then?

For to answer the important bit. 😛So what we gain is that we can build a dynamic list over an array and expose the _same_ list/array publicly in a read-only manner without any additional allocation.

**Comment by atifaziz** [comment]
Any further thoughts on this or did I miss some action/comment pending on my end?

**Comment by sebastienros** [comment]
@AmateurMemetics I assume it will also impact your PR in Jint
